# 📊Project: Spotify Streaming History Analysis | Data Cleaning

## Introduction

In this project, i will take the role of a Data Analyst to explore a real-world datasets from **Spotify Streaming History**. The main objectives are to **clean the data**, **handle null values**, **update data types**, and **detect anomalies** in the data before moving on to visualization.

In [1]:
#importing modules
import numpy as np
import pandas as pd
from datetime import datetime as dt
from IPython.display import FileLink

## Table of Contents (TOCs)

1. [Project](#)
    * [Introduction](#Introduction)
    * [Data Description](#Data-Description)
2. [Data Preparation](#Data-Preparation)
3. [Data Cleaning](#Data-Cleaning)
    * [Data Quality Report](#Data-Quality-Report)
    * [Handling Null Values](#Handling-Null-Values)
    * [Converting Data Types](#Converting-Data-Types)
    * [Anomaly Detection](#Anomaly-Detection)
4. [Dataset Download](#Dataset-Download)

## Data Description

- **spotify_track_uri**: Spotify URI that uniquely identifies each track in the form of "spotify:track:<base-62 string>"
- **ts**: Timestamp indicating when the track stopped playing in UTC (Coordinated Universal Time)
- **platform**: Platform used when streaming the track
- **ms_played**: Number of milliseconds the stream was played
- **track_name**:	Name of the track
- **artist_name**:	Name of the artist
- **album_name**:	Name of the album
- **reason_start**:	Why the track started
- **reason_end**:	Why the track ended
- **shuffle**:	TRUE or FALSE depending on if shuffle mode was used when playing the track
- **skipped**:	TRUE of FALSE depending on if the user skipped to the next song



## Data Preparation

- **Spotify Streaming History**: https://mavenanalytics.io/data-playground/spotify-streaming-history

(The dataset is from 7/8/2013 to 12/15/2024)

In [2]:
spotify_history = pd.read_csv("C:/Users/dangl/OneDrive/Tài liệu/DATASET/Spotify+Streaming+History/spotify_history.csv")

In [3]:
display(spotify_history)

spotify_track_uri                   ts    platform  ms_played  \
0       2J3n32GeLmMjwuAzyhcSNe  2013-07-08 02:44:34  web player       3185   
1       1oHxIPqJyvAYHy0PVrDU98  2013-07-08 02:45:37  web player      61865   
2       487OPlneJNni3NWC8SYqhW  2013-07-08 02:50:24  web player     285386   
3       5IyblF777jLZj1vGHG2UD3  2013-07-08 02:52:40  web player     134022   
4       0GgAAB0ZMllFhbNc3mAodO  2013-07-08 03:17:52  web player          0   
...                        ...                  ...         ...        ...   
149855  4Fz1WWr5o0OrlIcZxcyZtK  2024-12-15 23:06:19     android       1247   
149856  0qHMhBZqYb99yhX9BHcIkV  2024-12-15 23:06:21     android       1515   
149857  0HHdujGjOZChTrl8lJWEIq  2024-12-15 23:06:22     android       1283   
149858  7peh6LUcdNPcMdrSH4JPsM  2024-12-15 23:06:23     android       1306   
149859  6iGU74CwXuT4XVepjc9Emf  2024-12-15 23:06:25     android       1893   

                                               track_name        artist_name  \
0                                     Say It, Just Say It       The Mowgli's   
1           Drinking from the Bottle (feat. Tinie Tempah)      Calvin Harris   
2                                             Born To Die       Lana Del Rey   
3                                        Off To The Races       Lana Del Rey   
4                                               Half Mast  Empire Of The Sun   
...                                                   ...                ...   
149855                                    On The Way Home         John Mayer   
149856             Magical Mystery Tour - Remastered 2009        The Beatles   
149857  Stop This Train - Live at the Nokia Theatre, L...         John Mayer   
149858             I Don't Trust Myself (With Loving You)         John Mayer   
149859                              God Only Knows - Mono     The Beach Boys   

                                               album_name reason_start  \
0                                    Waiting For The Dawn     autoplay   
1                                               18 Months     clickrow   
2                      Born To Die - The Paradise Edition     clickrow   
3                      Born To Die - The Paradise Edition    trackdone   
4                                      Walking On A Dream     clickrow   
...                                                   ...          ...   
149855                                    Paradise Valley       fwdbtn   
149856                               Magical Mystery Tour       fwdbtn   
149857  Where the Light Is: John Mayer Live In Los Ang...       fwdbtn   
149858                                          Continuum       fwdbtn   
149859                                         Pet Sounds       fwdbtn   

       reason_end  shuffle  skipped  
0        clickrow    False    False  
1        clickrow    False    False  
2         unknown    False    False  
3        clickrow    False    False  
4         nextbtn    False    False  
...           ...      ...      ...  
149855     fwdbtn     True     True  
149856     fwdbtn     True     True  
149857     fwdbtn     True     True  
149858     fwdbtn     True     True  
149859     fwdbtn     True     True  

[149860 rows x 11 columns]

## Data Cleaning

### Data Quality Report

In [4]:
def check_data_quality(df):
    quality_report = {
        'null_values': df.isnull().sum().to_dict(),
        'duplicate_values': df.duplicated().sum(),
        'info_values': df.info()
    }
    return quality_report

In [5]:
display(check_data_quality(spotify_history))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149860 entries, 0 to 149859
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   spotify_track_uri  149860 non-null  object
 1   ts                 149860 non-null  object
 2   platform           149860 non-null  object
 3   ms_played          149860 non-null  int64 
 4   track_name         149860 non-null  object
 5   artist_name        149860 non-null  object
 6   album_name         149860 non-null  object
 7   reason_start       149717 non-null  object
 8   reason_end         149743 non-null  object
 9   shuffle            149860 non-null  bool  
 10  skipped            149860 non-null  bool  
dtypes: bool(2), int64(1), object(8)
memory usage: 10.6+ MB


{'null_values': {'spotify_track_uri': 0,
  'ts': 0,
  'platform': 0,
  'ms_played': 0,
  'track_name': 0,
  'artist_name': 0,
  'album_name': 0,
  'reason_start': 143,
  'reason_end': 117,
  'shuffle': 0,
  'skipped': 0},
 'duplicate_values': np.int64(1185),
 'info_values': None}

In [6]:
# Percentage of null values
pct_null_values = (spotify_history.isnull().sum() * 100 / len(spotify_history)).sort_values(ascending = False)
pct_null_values.loc[pct_null_values > 0]

reason_start    0.095422
reason_end      0.078073
dtype: float64

Let's take a look:

- The dataset includes **149,860** rows and **11** columns. There are **1185** duplicate rows and several columns contains **null values**:

+) Data profiling of the Spotify History dataset reveals high integrity. Missing values are confined to **'reason_start'** (0.09%) and **'reason_end'** (0.08%).

+) Given the **negligible volume**, these records can be safely imputed without biasing the results

### Removing duplicate rows

In [7]:
# Check count before
before_count = len(spotify_history)
print(f"Rows BEFORE removing duplicates: {before_count}")

# Perform the drop
spotify_history = spotify_history.drop_duplicates()

# Check count after
after_count = len(spotify_history)
print(f"Rows AFTER removing duplicates:  {after_count}")

# Calculate the difference
print(f"Total rows deleted: {before_count - after_count}")

Rows BEFORE removing duplicates: 149860
Rows AFTER removing duplicates:  148675
Total rows deleted: 1185


### Handling Null Values

In [8]:
# Replace nulls with 'unknown' in the reason_start and reason_end columns
spotify_history['reason_start'] = spotify_history['reason_start'].fillna('unknown')
spotify_history['reason_end'] = spotify_history['reason_end'].fillna('unknown')

### Converting Data Types

In [9]:
spotify_history['ts'] = pd.to_datetime(spotify_history['ts'], format ="%Y-%m-%d %H:%M:%S")


In [10]:
def converting_data_type(dataset, col_name,dtypes):
    dataset[col_name] = dataset[col_name].astype(dtype)

In [11]:
spotify_history.info()

<class 'pandas.core.frame.DataFrame'>
Index: 148675 entries, 0 to 149859
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   spotify_track_uri  148675 non-null  object        
 1   ts                 148675 non-null  datetime64[ns]
 2   platform           148675 non-null  object        
 3   ms_played          148675 non-null  int64         
 4   track_name         148675 non-null  object        
 5   artist_name        148675 non-null  object        
 6   album_name         148675 non-null  object        
 7   reason_start       148675 non-null  object        
 8   reason_end         148675 non-null  object        
 9   shuffle            148675 non-null  bool          
 10  skipped            148675 non-null  bool          
dtypes: bool(2), datetime64[ns](1), int64(1), object(7)
memory usage: 11.6+ MB


## Anomaly Detection

### A. Check whether ms_played < 0?

In [12]:
# Returns the total number of rows where ms_played is negative
negative_count = (spotify_history['ms_played'] < 0).sum()
print(f"Number of negative values: {negative_count}")

Number of negative values: 0


### B.Check whether ms_played = 0 when reason_end = trackdone. 

In [13]:
# 1. Capture the count before the specific removal
before_count = len(spotify_history)

# 2. Filter out ONLY the rows where BOTH conditions are met
# We use the tilde (~) symbol to say "keep everything EXCEPT this specific combination"
spotify_history = spotify_history[~((spotify_history['ms_played'] == 0) & (spotify_history['reason_end'] == 'trackdone'))]

# 3. Calculate how many were removed
after_count = len(spotify_history)
removed_count = before_count - after_count

print(f"Rows before specific filter: {before_count}")
print(f"Rows after specific filter:  {after_count}")
print(f"Removed {removed_count} rows where ms_played was 0 with 'trackdone' status.")

Rows before specific filter: 148675
Rows after specific filter:  148660
Removed 15 rows where ms_played was 0 with 'trackdone' status.


### C. Check rows where ms_played > 1 min, but still labelled in skipped as True

In [14]:
# 1. Identify the inconsistency: skipped is True AND played > 1 minute (60,000 ms)
inconsistent_mask = (spotify_history['skipped'] == True) & (spotify_history['ms_played'] > 60000)

# 2. Print the number of records we are about to fix
print(f"Fixing {inconsistent_mask.sum()} inconsistent skip records...")

# 3. Use .loc to change 'skipped' from True to False for those specific rows
spotify_history.loc[inconsistent_mask, 'skipped'] = False

# 4. Verify the fix
still_inconsistent = ((spotify_history['skipped'] == True) & (spotify_history['ms_played'] > 60000)).sum()
print(f"Inconsistent records remaining: {still_inconsistent}")

Fixing 1316 inconsistent skip records...
Inconsistent records remaining: 0


## Dataset Download

In [15]:
spotify_history.to_csv('spotify_cleaned.csv', index=False)

In [16]:
# This generates a blue link in the output that you can click to download
FileLink('spotify_cleaned.csv')

C:\Users\dangl\spotify_cleaned.csv